src/main/java/emu/grasscutter/game/dungeons/DungeonSystem.java

```diff
@@ -35,128 +36,131 @@ public DungeonSystem(GameServer server) {
-        return true;
+        return enterDungeon(player, pointId, dungeonId, BASIC_DUNGEON_SETTLE_LISTENER);
     }
 
     /**
-     * used in tower dungeons handoff
-     */
-    public boolean handoffDungeon(Player player, int dungeonId, List<DungeonSettleListener> dungeonSettleListeners) {
-        DungeonData data = GameData.getDungeonDataMap().get(dungeonId);
-
+     * TODO record also player's position for dungeon that does not have entries and exits,
+     * like trial avatar activity and mist trial activity
+     * */
+    public boolean enterDungeon(Player player, int pointId, int dungeonId, DungeonSettleListener dungeonSettleListeners) {
+        val data = GameData.getDungeonDataMap().get(dungeonId);
         if (data == null) {
+            Grasscutter.getLogger().error("No resource found for this dungeon: {}", dungeonId);
             return false;
         }
-        Grasscutter.getLogger().info("{}({}) is trying to enter tower dungeon {}" ,player.getNickname(),player.getUid(),dungeonId);
 
+        Grasscutter.getLogger().info("{}({}) is trying to enter {}({})", player.getNickname(), player.getUid(), data.getType(), dungeonId);
+        player.getScene().setPrevScene(player.getSceneId());
         if (player.getWorld().transferPlayerToScene(player, data.getSceneId(), data)) {
-            dungeonSettleListeners.forEach(player.getScene()::addDungeonSettleObserver);
+            player.getScene().setDungeonManager(new DungeonManager(player.getScene(), data));
+            player.getScene().addDungeonSettleObserver(dungeonSettleListeners);
         }
+
+        player.getScene().setPrevScenePoint(Optional.ofNullable(GameData.getDungeonEntriesMap().get(dungeonId))
+            .map(DungeonEntries::getEntryPoint).map(PointData::getId).orElse(pointId));
         return true;
     }
 
-    public void exitDungeon(Player player) {
-        Scene scene = player.getScene();
-
-        if (scene==null || scene.getSceneType() != SceneType.SCENE_DUNGEON) {
-            return;
-        }
+    // TODO check, modify it to work on multiplayer
+    public void restartDungeon(Player player) {
+        restartDungeon(player, BASIC_DUNGEON_SETTLE_LISTENER);
+    }
 
-        // Get previous scene
-        int prevScene = scene.getPrevScene() > 0 ? scene.getPrevScene() : 3;
+    public void restartDungeon(Player player, DungeonSettleListener listener) {
+        val scene = player.getScene();
+        if (scene == null || scene.getDungeonManager() == null) return;
 
-        // Get previous position
-        val dungeonManager = scene.getDungeonManager();
-        DungeonData dungeonData =  dungeonManager != null ? dungeonManager.getDungeonData() : null;
-        Position prevPos = new Position(GameConstants.START_POSITION);
+        scene.getScriptManager().onDestroy();
+        scene.getWorld().deregisterScene(scene);
+        enterDungeon(player, 0, scene.getDungeonManager().getDungeonData().getId(), listener);
```

Comment:
Yes, thats why i made a pre made a map to record the dungeon id as the key and the entry and exits point as the value to counter this issue. Cuz there are cases where we will be moving from one dungeon to another, not just entering from scene 3 and exit, if we use the old logic, we will be stuck between the dungeons if we move from one dungeon to another (since the dungeon scene id is always gonna be larger than 0). The pre-made map will tell where we need to go when we are exiting from the dungeon regardless of where we are entering.

Reply:
Yep, but wasn't there also mentions of dungeons that have no output point and will normally warp the user back to its last location in your code comment?
Those would need some way to persevere the info. Through I guess it's a todo anyway, so it might be fine for now.

Reply:
yea for now im just having the user to warp back to the default born position for those that doesn't have exit points, so that have to be fixed later